extension/icu/icu-timebucket.cpp

```diff
@@ -0,0 +1,725 @@
+	D_ASSERT(args.ColumnCount() == 2);
+
+	auto &func_expr = (BoundFunctionExpression &)state.expr;
+	auto &info = (BindData &)*func_expr.bind_info;
+	CalendarPtr calendar_ptr(info.calendar->clone());
+	auto calendar = calendar_ptr.get();
+	SetTimeZone(calendar, string_t("UTC"));
+
+	auto &bucket_width_arg = args.data[0];
+	auto &ts_arg = args.data[1];
+
+	if (bucket_width_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		if (ConstantVector::IsNull(bucket_width_arg)) {
+			result.SetVectorType(VectorType::CONSTANT_VECTOR);
+			ConstantVector::SetNull(result, true);
+		} else {
+			interval_t bucket_width = *ConstantVector::GetData<interval_t>(bucket_width_arg);
+			if (bucket_width.months == 0) {
+				int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+				if (bucket_width_micros <= 0) {
+					throw NotImplementedException("Period must be greater than 0");
+				}
+				BinaryExecutor::Execute<interval_t, timestamp_t, timestamp_t>(
+				    bucket_width_arg, ts_arg, result, args.size(), [&](interval_t bucket_width, timestamp_t ts) {
+					    return WidthLessThanDaysBinaryOperator::Operation<interval_t, timestamp_t, timestamp_t>(
+					        bucket_width, ts, calendar);
+				    });
+			} else if (bucket_width.months != 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+				if (bucket_width.months < 0) {
+					throw NotImplementedException("Period must be greater than 0");
+				}
+				BinaryExecutor::Execute<interval_t, timestamp_t, timestamp_t>(
+				    bucket_width_arg, ts_arg, result, args.size(), [&](interval_t bucket_width, timestamp_t ts) {
+					    return WidthMoreThanMonthsBinaryOperator::Operation<interval_t, timestamp_t, timestamp_t>(
+					        bucket_width, ts, calendar);
+				    });
+			} else {
+				throw NotImplementedException("Month intervals cannot have day or time component");
+			}
+		}
+	} else {
+		BinaryExecutor::Execute<interval_t, timestamp_t, timestamp_t>(
+		    bucket_width_arg, ts_arg, result, args.size(), [&](interval_t bucket_width, timestamp_t ts) {
+			    return BinaryOperator::Operation<interval_t, timestamp_t, timestamp_t>(bucket_width, ts, calendar);
+		    });
+	}
+}
+
+void ICUTimeBucket::ICUTimeBucketOffsetFunction(DataChunk &args,
+                                                                                      ExpressionState &state,
+                                                                                      Vector &result) {
+	D_ASSERT(args.ColumnCount() == 3);
+
+	auto &func_expr = (BoundFunctionExpression &)state.expr;
+	auto &info = (BindData &)*func_expr.bind_info;
+	CalendarPtr calendar_ptr(info.calendar->clone());
+	auto calendar = calendar_ptr.get();
+	SetTimeZone(calendar, string_t("UTC"));
+
+	auto &bucket_width_arg = args.data[0];
+	auto &ts_arg = args.data[1];
+	auto &offset_arg = args.data[2];
+
+	if (bucket_width_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
```

Comment:
`TernaryExecutor::Execute` will do all the null checking for you. Or is the point here to throw even if the other arguments are `NULL`?

Reply:
I'll use `TernaryExecutor::Execute` instead of throwing error.